src/Language/PureScript/TypeChecker/Deriving.hs

```diff
@@ -359,6 +368,128 @@ decomposeRec' = sortOn fst . go
+          consHintData (Right lbl) $
+            fmap (lbl, ) <$> go ty
+
+      TypeApp _ tyFn tyArg -> do
+        assertNoParamUsedIn tyFn
+        fmap MentionsParam <$> go tyArg
+
+      TypeVar _ name ->
+        pure $ (name == param) `orEmpty` IsParam
+
+      ty ->
+        assertNoParamUsedIn ty $> Nothing
+
+usingLamIdent :: forall m. MonadSupply m => (Expr -> m Expr) -> m Expr
+usingLamIdent cb = do
+  ident <- freshIdent "v"
+  lam ident <$> cb (mkVar ident)
+
+traverseFields :: forall f. Applicative f => (ParamUsage -> Expr -> f Expr) -> NonEmpty (PSString, ParamUsage) -> Expr -> f Expr
+traverseFields f fields r = fmap (ObjectUpdate r) . for (toList fields) $ \(lbl, usage) -> (lbl, ) <$> f usage (Accessor lbl r)
+
+unnestRecords :: forall f. Applicative f => (ParamUsage -> Expr -> f Expr) -> ParamUsage -> Expr -> f Expr
+unnestRecords f = fix $ \go -> \case
+  IsRecord fields -> traverseFields go fields
+  usage -> f usage
+
+mkCasesForTraversal
+  :: forall f m
+   . Applicative f -- this effect distinguishes the semantics of maps, folds, and traversals
+  => MonadSupply m
+  => ModuleName
+  -> (ParamUsage -> Expr -> f Expr) -- how to handle constructor arguments
+  -> (f Expr -> m Expr) -- resolve the applicative effect into an expression
+  -> [(ProperName 'ConstructorName, [Maybe ParamUsage])]
+  -> m Expr
+mkCasesForTraversal mn handleArg extractExpr ctors = do
+  m <- freshIdent "m"
+  fmap (lamCase m) . for ctors $ \(ctorName, ctorUsages) -> do
+    ctorArgs <- for ctorUsages $ \usage -> freshIdent "v" <&> (, usage)
+    let ctor = mkCtor mn ctorName
+    let caseBinder = mkCtorBinder mn ctorName $ map (mkBinder . fst) ctorArgs
+    fmap (CaseAlternative [caseBinder] . unguarded) . extractExpr $
+      fmap (foldl' App ctor) . for ctorArgs $ \(ident, mbUsage) -> maybe pure handleArg mbUsage $ mkVar ident
+
+data TraversalOps m = forall f. Applicative f => TraversalOps
+  { visitExpr :: m Expr -> f Expr -- lift an expression into the applicative effect defining the traversal
+  , extractExpr :: f Expr -> m Expr -- resolve the applicative effect into an expression
+  }
+
+mkTraversal
+  :: forall m
+   . MonadSupply m
+  => ModuleName
+  -> Expr -- a var representing map, foldMap, or traverse, for handling structured values
+  -> TraversalOps m
+  -> [(ProperName 'ConstructorName, [Maybe ParamUsage])]
+  -> m Expr
+mkTraversal mn recurseVar (TraversalOps @_ @f visitExpr extractExpr) ctors = do
+  f <- freshIdent "f"
+  let
+    handleValue :: ParamUsage -> Expr -> f Expr
+    handleValue = unnestRecords $ \usage inputExpr -> visitExpr $ flip App inputExpr <$> mkFnExprForValue usage
+
+    mkFnExprForValue :: ParamUsage -> m Expr
+    mkFnExprForValue = \case
+      IsParam ->
+        pure $ mkVar f
+      MentionsParam innerUsage ->
+        App recurseVar <$> mkFnExprForValue innerUsage
+      IsRecord fields ->
+        usingLamIdent $ extractExpr . traverseFields handleValue fields
```

Comment:
Just checking my understanding. Because `unnestRecords` is being used and it always consumes an `IsRecord`, doesn't that mean this code path is never triggered? Meaning, this code exists only to satisfy type safety, not because it does anything useful in this current state.

Reply:
No, it does something useful if `mkFnExprForValue` is called with a `MentionsParam (IsRecord _)`, for example. `unnestRecords` doesn't recurse into the `ParamUsage` it's given, which is what we want because if a constructor argument is an `Array { foo :: a }`, we can't unnest the `foo`s in the current lambda.

Reply:
Ah, ok!